### PR TITLE
[WIP] Feat/sitemaps extension

### PR DESCRIPTION
**Pull request by tuxnco**
Provide support for google extensions to the sitemaps protocol in XML sitemaps.

Supporting those through the mRSS can be done in an other PR.

### CODE REVIEW

**Comment by tuxnco** [comment]
Need to add support for validation / error reporting to make something really useful.

File: src/main/java/crawlercommons/sitemaps/ImageAttributes.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    private ImageAttributes(){}
+
+    public ImageAttributes(URL loc) {
+        this.loc = loc;
+    }
+
+    @Override
+    public boolean equals(Object other) {
```
Comment (panthony) [inline src/main/java/crawlercommons/sitemaps/ImageAttributes.java:100]: It's dangerous to override `equals` but not `hashCode`. A user may have unexpected behevior if using this object in a map, etc.

File: src/main/java/crawlercommons/sitemaps/ImageAttributes.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        if (other == null) {
+            return false;
+        }
+        if (other instanceof ImageAttributes) {
```
Comment (panthony) [inline src/main/java/crawlercommons/sitemaps/ImageAttributes.java]: I think I would do early returns in this case.

```java
if (!(other instanceof ImageAttributes)) {
return false;
}

if (!Objects.equals(loc, that.loc)) {
 return false;
}
```

File: src/main/java/crawlercommons/sitemaps/LinkAttributes.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    public Map<String, String> getParams() {
+        return params;
+    }
+
+    public void setParams(Map<String, String> params) {
+        this.params = params;
+    }
+
+    @Override
+    public boolean equals(Object other) {
```
Comment (panthony) [inline src/main/java/crawlercommons/sitemaps/LinkAttributes.java:60]: same as ImageAttributes

File: src/main/java/crawlercommons/sitemaps/NewsAttributes.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        return stockTickers;
+    }
+
+    public void setStockTickers(String[] stockTickers) {
+        this.stockTickers = stockTickers;
+    }
+
+
+    @Override
+    public boolean equals(Object other) {
```
Comment (panthony) [inline src/main/java/crawlercommons/sitemaps/NewsAttributes.java:131]: same as ImageAttributes

File: src/main/java/crawlercommons/sitemaps/SiteMapURL.java [hunk 5]
```diff
@@ -68,6 +67,30 @@
      */
     private boolean valid;
 
+    public ImageAttributes[] getImages() {
```
Comment (panthony) [inline src/main/java/crawlercommons/sitemaps/SiteMapURL.java]: Should be moved with other accessors

File: src/main/java/crawlercommons/sitemaps/SiteMapURL.java [hunk 6]
```diff
@@ -290,6 +383,26 @@ public boolean equals(Object o) {
...
+
+        if (!Arrays.deepEquals(images, that.images)) {
+            return false;
+        }
+
+        if (!Arrays.deepEquals(links, that.links)) {
+            return false;
+        }
+
+        if (news == null) {
```
Comment (panthony) [inline src/main/java/crawlercommons/sitemaps/SiteMapURL.java]: Since java7 there is `Objects.equals` that handles null values.

Comment (tuxnco) [reply src/main/java/crawlercommons/sitemaps/SiteMapURL.java]: Much appreciated!

File: src/main/java/crawlercommons/sitemaps/ImageAttributes.java (hunk 1 above)
Comment (tuxnco) [reply src/main/java/crawlercommons/sitemaps/ImageAttributes.java:100]: Correct, I fixed that

**Pull request merged**